src/archivist/recorder/repositories/git/index.js

```diff
@@ -0,0 +1,203 @@
+/**
+ * This module is the boundary beyond which the usage of git is abstracted.
+ * Commit SHAs are used as opaque unique IDs.
+ */
+
+import fsApi from 'fs';
+import path from 'path';
+
+import mime from 'mime';
+
+import RepositoryInterface from '../interface.js';
+
+import DataMapper, { COMMIT_MESSAGE_PREFIXES_REGEXP, COMMIT_MESSAGE_PREFIX } from './dataMapper.js';
+import Git from './git.js';
+
+const fs = fsApi.promises;
+const PDF_MIME_TYPE = 'application/pdf';
+
+mime.define({ 'text/markdown': ['md'] }, true); // ensure extension for markdown files is `.md` and not `.markdown`
+
+export default class GitRepository extends RepositoryInterface {
+  constructor({ path, author, publish, prefixMessageToSnapshotId }) {
+    super();
+    this.path = path;
+    this.needsPublication = publish;
+    this.git = new Git({ path: this.path, author });
+    this.prefixMessageToSnapshotId = prefixMessageToSnapshotId;
+  }
+
+  async initialize() {
+    await this.git.initialize();
+
+    return this;
+  }
+
+  async save(record) {
+    const { serviceId, documentType, fetchDate } = record;
+
+    if (record.isFirstRecord === undefined || record.isFirstRecord === null) {
+      record.isFirstRecord = await this.#isFirstRecord(serviceId, documentType);
+    }
+    const { message, content, fileExtension } = await this.#toPersistence(record);
+
+    const filePath = await this.#writeFile({ serviceId, documentType, content, fileExtension });
+    const sha = await this.#commit({ filePath, message, date: fetchDate });
+
+    if (!sha) {
+      return Object(null);
+    }
+
+    record.id = sha;
+
+    return record;
+  }
+
+  finalize() {
+    if (!this.needsPublication) {
+      return;
+    }
+
+    return this.git.pushChanges();
```

Comment:
That method is `async`, how can this one not be?

Reply:
When marking method `async` it will just wrap it in a promise. In this case I return a promise, so there is no need to wrap it.